src/trace/trace.c

```diff
@@ -96,19 +115,120 @@ static inline void mtrace_event(const char *data, uint32_t length)
 	platform_shared_commit(trace, sizeof(*trace));
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	STATIC_ASSERT(LOG_LEVEL_CRITICAL < LOG_LEVEL_VERBOSE,
 		      LOG_LEVEL_CRITICAL_MUST_HAVE_LOWEST_VALUE);
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+static void emit_recent_entry(struct recent_log_entry *entry)
+{
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %pE",
```

Comment:
This is interesting. Has this been decided? This is a new format, with it you'll get something like
```
Suppressed 2 similar messages: perf comp_copy peak plat %d cpu %d
```
I think we're trying to comply with the standard for printing formats. If not with POSIX then at least with the kernel. In the kernel "%pE" is used for "printing raw buffer as an escaped string." This isn't the same as that, right? But I also don't see a suitable format, so, I'm not sure what a good choice here be. We also could in principle consider printing messages like
```
Suppressed 2 similar messages: function _rfree_unlocked() line 925
```
or "file alloc.c line 925." Not sure either of these options is better, but at least we could have standard format strings for them. E.g. if we use function + line number, we could use some form of a %ps / %pS format. @lgirdwood ?